### PR TITLE
[CI] Add Clang-Tidy and Cpplint checks

### PR DESCRIPTION
**Pull request by feltech**
Add `clang-tidy` and `cpplint` invocations as part of the CMake build process, both in CI and (optionally) locally.

CMake boolean variables `OPENASSETIO_ENABLE_CLANG_TIDY` and `OPENASSETIO_ENABLE_CPPLINT` are introduced to allow enabling these (disabled by default).

The Clang-Tidy style is lifted and modified from GCP's repo: https://github.com/googleapis/google-cloud-cpp

### CODE REVIEW

**Comment by feltech** [comment]
For testing the linters: the final commit d1d1bb35 (`fixup! [CI] Integrate C++ linters into CI`) is deliberately split out so that linter checks can be tested by undoing one of the fixes in there (see commit message).

File: .clang-tidy [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+CheckOptions:
+    - { key: readability-identifier-naming.NamespaceCase,          value: camelBack  }
+    - { key: readability-identifier-naming.ClassCase,              value: CamelCase  }
+    - { key: readability-identifier-naming.StructCase,             value: CamelCase  }
+    - { key: readability-identifier-naming.TemplateParameterCase,  value: CamelCase  }
+    - { key: readability-identifier-naming.FunctionCase,           value: camelBack  }
+    - { key: readability-identifier-naming.VariableCase,           value: camelBack  }
+    - { key: readability-identifier-naming.ClassMemberCase,        value: camelBack  }
+    - { key: readability-identifier-naming.ClassMemberSuffix,      value: _          }
```
Comment (foundrytom) [inline .clang-tidy:65]: Not sure about this if it includes the public API.

File: .clang-tidy [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+#
+#  -modernize-return-braced-init-list: We think removing typenames and using
+#      only braced-init can hurt readability.
+#
+#  -modernize-avoid-c-arrays: We only use C arrays when they seem to be the
+#      right tool for the job, such as `char foo[] = "hello"`. In these cases,
+#      avoiding C arrays often makes the code less readable, and std::array is
+#      not a drop-in replacement because it doesn't deduce the size.
+#
+#  -google-readability-braces-around-statements: This is a duplicate of
```
Comment (foundrytom) [inline .clang-tidy]: [fix] Can we re-enable this please! Its just asking for trouble :)

Comment (feltech) [reply .clang-tidy]: Updated in fca16de4

File: .clang-tidy (hunk 1 above)
Comment (feltech) [reply .clang-tidy]: Updated in fca16de4

**Review by carmenpinto — APPROVED** [review]

File: .clang-tidy [hunk 3]
```diff
@@ -0,0 +1,73 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright 2022 Google LLC
+# SPDX-License-Identifier: Apache-2.0
```
Comment (foundrytom) [inline .clang-tidy:3]: [fix] Can we remove the duplicate license identifier

**Comment by foundrytom** [comment]
@feltech Feel free to squash and merge once that duplicate lic is fixed

File: .clang-tidy (hunk 3 above)
Comment (feltech) [reply .clang-tidy:3]: Will do.  I wasn't sure how it should be listed, e.g. if each copyright needs to be paired with a license, but then I guess multiple copyrights only makes sense if they're all Apache 2?

Comment (foundrytom) [reply .clang-tidy]: Sounds right - as you can't assume that you can mix licenses.

**Pull request merged**